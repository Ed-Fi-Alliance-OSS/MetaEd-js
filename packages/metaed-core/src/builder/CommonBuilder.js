// @flow
import { MetaEdGrammar } from '../grammar/gen/MetaEdGrammar';
import { TopLevelEntityBuilder } from './TopLevelEntityBuilder';
import { newCommon, newInlineCommon } from '../model/Common';
import { isErrorText } from './BuilderUtility';
import { NoTopLevelEntity } from '../model/TopLevelEntity';
import { sourceMapFrom } from '../model/SourceMap';

export class CommonBuilder extends TopLevelEntityBuilder {
  enterCommon(context: MetaEdGrammar.CommonContext) {
    this.enteringEntity(newCommon);
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      this.currentTopLevelEntity.sourceMap.type = sourceMapFrom(context);
    }
  }

  // eslint-disable-next-line no-unused-vars
  exitCommon(context: MetaEdGrammar.CommonContext) {
    this.exitingEntity();
  }

  enterCommonName(context: MetaEdGrammar.CommonNameContext) {
    if (this.currentTopLevelEntity === NoTopLevelEntity) return;
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.enteringName(context.ID().getText());
    this.currentTopLevelEntity.sourceMap.metaEdName = sourceMapFrom(context);
  }

  enterInlineCommon(context: MetaEdGrammar.InlineCommonContext) {
    this.enteringEntity(newInlineCommon);
    if (this.currentTopLevelEntity !== NoTopLevelEntity) {
      this.currentTopLevelEntity.sourceMap.type = sourceMapFrom(context);
    }
  }

  // eslint-disable-next-line no-unused-vars
  exitInlineCommon(context: MetaEdGrammar.InlineCommonContext) {
    this.exitingEntity();
  }

  enterInlineCommonName(context: MetaEdGrammar.InlineCommonNameContext) {
    if (context.exception || context.ID() == null || context.ID().exception || isErrorText(context.ID().getText())) return;
    this.enteringName(context.ID().getText());
    this.currentTopLevelEntity.sourceMap.metaEdName = sourceMapFrom(context);
  }
}
