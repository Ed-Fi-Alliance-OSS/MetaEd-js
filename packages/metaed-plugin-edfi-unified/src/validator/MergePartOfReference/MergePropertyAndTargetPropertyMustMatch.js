// @flow
import type { EntityProperty, MetaEdEnvironment, PropertyType, ValidationFailure } from 'metaed-core';
import { asReferentialProperty, asModelType, getPropertiesOfType } from 'metaed-core';
import {
  findReferencedProperty,
  getReferencedEntity,
  referenceTypes,
  matchAllButFirstAsIdentityProperties,
  matchAllIdentityReferenceProperties,
} from '../ValidatorShared/FindReferencedProperty';

const validPropertyTypes: Array<PropertyType> = ['association', 'choice', 'common', 'domainEntity', 'inlineCommon'];

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  getPropertiesOfType(metaEd.propertyIndex, ...validPropertyTypes).forEach(property => {
    if (!property.mergedProperties || property.mergedProperties.length === 0) return;

    asReferentialProperty(property).mergedProperties.forEach(mergedProperty => {
      const mergeProperty: ?EntityProperty = findReferencedProperty(
        metaEd.entity,
        property.parentEntity,
        mergedProperty.mergePropertyPath,
        matchAllButFirstAsIdentityProperties(),
      );
      const targetProperty: ?EntityProperty = findReferencedProperty(
        metaEd.entity,
        property.parentEntity,
        mergedProperty.targetPropertyPath,
        matchAllIdentityReferenceProperties(),
      );

      if (!mergeProperty || !targetProperty) return;

      if (mergeProperty && targetProperty) {
        if (mergeProperty.type === targetProperty.type) {
          if (mergeProperty.metaEdName === targetProperty.metaEdName) return;

          if (
            referenceTypes.includes(asModelType(mergeProperty.type)) &&
            referenceTypes.includes(asModelType(targetProperty.type))
          ) {
            const mergeBaseEntity = getReferencedEntity(metaEd.entity, mergeProperty.metaEdName, mergeProperty.type);
            const targetBaseEntity = getReferencedEntity(metaEd.entity, targetProperty.metaEdName, targetProperty.type);

            if (mergeBaseEntity && mergeBaseEntity.baseEntityName === targetProperty.metaEdName) return;
            if (targetBaseEntity && targetBaseEntity.baseEntityName === mergeProperty.metaEdName) return;
          }
        }
      }

      failures.push({
        validatorName: 'MergePropertyAndTargetPropertyMustMatch',
        category: 'error',
        message: `The merge paths '${mergedProperty.mergePropertyPath.join(
          '.',
        )}' and '${mergedProperty.targetPropertyPath.join('.')}' do not correspond to the same entity name and/or type.`,
        sourceMap: mergedProperty.sourceMap.mergePropertyPath[0],
        fileMap: null,
      });
    });
  });

  return failures;
}
