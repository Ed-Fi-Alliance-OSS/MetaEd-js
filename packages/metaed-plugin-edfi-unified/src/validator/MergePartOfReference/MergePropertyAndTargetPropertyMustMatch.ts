import { EntityProperty, MetaEdEnvironment, PropertyType, ValidationFailure, ModelBase } from 'metaed-core';
import {
  asReferentialProperty,
  asModelType,
  getPropertiesOfType,
  asTopLevelEntity,
  isReferentialProperty,
} from 'metaed-core';
import {
  findReferencedProperty,
  getReferencedEntities,
  referenceTypes,
  matchAllButFirstAsIdentityProperties,
  matchAllIdentityReferenceProperties,
} from '../ValidatorShared/FindReferencedProperty';

const validPropertyTypes: Array<PropertyType> = [
  'association',
  'choice',
  'common',
  'descriptor',
  'domainEntity',
  'enumeration',
  'inlineCommon',
  'schoolYearEnumeration',
  'sharedDecimal',
  'sharedInteger',
  'sharedShort',
  'sharedString',
];

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  getPropertiesOfType(metaEd.propertyIndex, ...validPropertyTypes).forEach(property => {
    if (!isReferentialProperty(property)) return;
    const referentialProperty = asReferentialProperty(property);
    if (referentialProperty.mergedProperties.length === 0) return;
    const { namespace } = referentialProperty;
    referentialProperty.mergedProperties.forEach(mergedProperty => {
      const mergeProperty: EntityProperty | null = findReferencedProperty(
        namespace,
        referentialProperty.parentEntity,
        mergedProperty.mergePropertyPath,
        matchAllButFirstAsIdentityProperties(),
      );
      const targetProperty: EntityProperty | null = findReferencedProperty(
        namespace,
        referentialProperty.parentEntity,
        mergedProperty.targetPropertyPath,
        matchAllIdentityReferenceProperties(),
      );

      if (!mergeProperty || !targetProperty) return;

      if (mergeProperty && targetProperty) {
        if (mergeProperty.type === targetProperty.type) {
          if (mergeProperty.metaEdName === targetProperty.metaEdName) return;

          if (
            referenceTypes.includes(asModelType(mergeProperty.type)) &&
            referenceTypes.includes(asModelType(targetProperty.type))
          ) {
            const mergeBaseEntity: Array<ModelBase> = getReferencedEntities(
              namespace,
              mergeProperty.metaEdName,
              mergeProperty.referencedNamespaceName,
              mergeProperty.type,
            );
            const targetBaseEntity: Array<ModelBase> = getReferencedEntities(
              namespace,
              targetProperty.metaEdName,
              targetProperty.referencedNamespaceName,
              targetProperty.type,
            );

            if (mergeBaseEntity[0] && asTopLevelEntity(mergeBaseEntity[0]).baseEntityName === targetProperty.metaEdName)
              return;
            if (targetBaseEntity[0] && asTopLevelEntity(targetBaseEntity[0]).baseEntityName === mergeProperty.metaEdName)
              return;
          }
        }
      }

      failures.push({
        validatorName: 'MergePropertyAndTargetPropertyMustMatch',
        category: 'error',
        message: `The merge paths '${mergedProperty.mergePropertyPath.join(
          '.',
        )}' and '${mergedProperty.targetPropertyPath.join('.')}' do not correspond to the same entity name and/or type.`,
        sourceMap: mergedProperty.sourceMap.mergePropertyPath[0],
        fileMap: null,
      });
    });
  });

  return failures;
}
