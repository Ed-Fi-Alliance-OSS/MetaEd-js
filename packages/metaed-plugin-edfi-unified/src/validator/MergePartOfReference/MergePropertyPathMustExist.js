// @flow
import R from 'ramda';
import type {
  PropertyType,
  MetaEdEnvironment,
  ValidationFailure,
} from '../../../../metaed-core/index';
import { getPropertiesOfType } from '../../../../metaed-core/index';
import { asReferentialProperty } from '../../../../metaed-core/src/model/property/ReferentialProperty';
import { failReferencedPropertyDoesNotExist } from '../ValidatorShared/FailReferencedPropertyDoesNotExist';

const validPropertyTypes: Array<PropertyType> = [
  'association',
  'choice',
  'common',
  'domainEntity',
  'inlineCommon',
];

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  getPropertiesOfType(metaEd.propertyIndex, ...validPropertyTypes).forEach(property => {
    if (!property.mergedProperties) return;
    asReferentialProperty(property).mergedProperties.forEach(mergedProperty => {
      failReferencedPropertyDoesNotExist(
        'MergePropertyPathMustExist',
        metaEd.entity,
        property.parentEntity,
        mergedProperty.mergePropertyPath,
        R.head(mergedProperty.targetPropertyPath),
        R.head(mergedProperty.sourceMap.mergePropertyPath),
        failures,
      );
    });
  });

  return failures;
}
