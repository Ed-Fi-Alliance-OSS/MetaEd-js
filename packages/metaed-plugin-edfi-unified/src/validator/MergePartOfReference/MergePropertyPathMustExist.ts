import R from 'ramda';
import { PropertyType, MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { getPropertiesOfType, asReferentialProperty } from 'metaed-core';
import { failReferencedPropertyDoesNotExist } from '../ValidatorShared/FailReferencedPropertyDoesNotExist';

const validPropertyTypes: Array<PropertyType> = [
  'association',
  'choice',
  'common',
  'descriptor',
  'domainEntity',
  'enumeration',
  'inlineCommon',
  'schoolYearEnumeration',
  'sharedDecimal',
  'sharedInteger',
  'sharedShort',
  'sharedString',
];

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  getPropertiesOfType(metaEd.propertyIndex, ...validPropertyTypes).forEach(property => {
    // TODO: As of METAED-881, the current property here could also be one of the shared simple properties, which
    // are not currently extensions of ReferentialProperty but have an equivalent mergedProperties field
    const referentialProperty = asReferentialProperty(property);
    referentialProperty.mergedProperties.forEach(mergedProperty => {
      failReferencedPropertyDoesNotExist(
        'MergePropertyPathMustExist',
        referentialProperty.namespace,
        referentialProperty.parentEntity,
        mergedProperty.mergePropertyPath,
        R.head(mergedProperty.targetPropertyPath),
        R.head(mergedProperty.sourceMap.mergePropertyPath),
        failures,
      );
    });
  });

  return failures;
}
