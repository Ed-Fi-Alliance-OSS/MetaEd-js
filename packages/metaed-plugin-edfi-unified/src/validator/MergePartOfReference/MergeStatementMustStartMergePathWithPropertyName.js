// @flow
import type { PropertyType, MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { getPropertiesOfType, asReferentialProperty, isReferentialProperty } from 'metaed-core';

const validTypes: Array<PropertyType> = ['common', 'inlineCommon', 'choice', 'association', 'domainEntity'];

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  getPropertiesOfType(metaEd.propertyIndex, ...validTypes).forEach(property => {
    if (!isReferentialProperty(property)) return;
    const referentialProperty = asReferentialProperty(property);
    referentialProperty.mergedProperties.forEach(mergedProperty => {
      const prefix: string =
        referentialProperty.withContext && referentialProperty.withContext !== referentialProperty.metaEdName
          ? referentialProperty.withContext
          : '';
      if (mergedProperty.mergePropertyPath[0] === `${prefix}${referentialProperty.metaEdName}`) return;

      failures.push({
        validatorName: 'MergeStatementMustStartMergePathWithPropertyName',
        category: 'error',
        message: `Merge statement must start first property path with the current property ${
          prefix ? 'context and ' : ''
        }name: ${prefix}${referentialProperty.metaEdName}.`,
        sourceMap: mergedProperty.sourceMap.mergePropertyPath[0],
        fileMap: null,
      });
    });
  });
  return failures;
}
