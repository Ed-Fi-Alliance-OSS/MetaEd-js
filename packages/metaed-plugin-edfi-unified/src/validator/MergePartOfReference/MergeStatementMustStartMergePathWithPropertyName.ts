import { PropertyType, MetaEdEnvironment, ValidationFailure } from 'metaed-core';
import { getPropertiesOfType, asReferentialProperty } from 'metaed-core';

const validPropertyTypes: Array<PropertyType> = [
  'association',
  'choice',
  'common',
  'descriptor',
  'domainEntity',
  'enumeration',
  'inlineCommon',
  'schoolYearEnumeration',
  'sharedDecimal',
  'sharedInteger',
  'sharedShort',
  'sharedString',
];
export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];
  getPropertiesOfType(metaEd.propertyIndex, ...validPropertyTypes).forEach(property => {
    // TODO: As of METAED-881, the current property here could also be one of the shared simple properties, which
    // are not currently extensions of ReferentialProperty but have an equivalent mergedProperties field
    const referentialProperty = asReferentialProperty(property);
    referentialProperty.mergedProperties.forEach(mergedProperty => {
      const prefix: string =
        referentialProperty.withContext && referentialProperty.withContext !== referentialProperty.metaEdName
          ? referentialProperty.withContext
          : '';
      if (mergedProperty.mergePropertyPath[0] === `${prefix}${referentialProperty.metaEdName}`) return;

      failures.push({
        validatorName: 'MergeStatementMustStartMergePathWithPropertyName',
        category: 'error',
        message: `Merge statement must start first property path with the current property ${
          prefix ? 'context and ' : ''
        }name: ${prefix}${referentialProperty.metaEdName}.`,
        sourceMap: mergedProperty.sourceMap.mergePropertyPath[0],
        fileMap: null,
      });
    });
  });
  return failures;
}
