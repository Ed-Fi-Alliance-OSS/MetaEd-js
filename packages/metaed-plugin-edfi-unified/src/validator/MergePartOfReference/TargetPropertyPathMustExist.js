// @flow
import R from 'ramda';
import type { PropertyType, MetaEdEnvironment, ValidationFailure, Namespace } from 'metaed-core';
import { getPropertiesOfType, asReferentialProperty, isReferentialProperty } from 'metaed-core';
import { failReferencedPropertyDoesNotExist } from '../ValidatorShared/FailReferencedPropertyDoesNotExist';

const validPropertyTypes: Array<PropertyType> = [
  'association',
  'choice',
  'common',
  'descriptor',
  'domainEntity',
  'enumeration',
  'inlineCommon',
  'schoolYearEnumeration',
];

export function validate(metaEd: MetaEdEnvironment): Array<ValidationFailure> {
  const failures: Array<ValidationFailure> = [];

  getPropertiesOfType(metaEd.propertyIndex, ...validPropertyTypes).forEach(property => {
    if (!isReferentialProperty(property)) return;
    const referentialProperty = asReferentialProperty(property);
    const namespaces: Array<Namespace> = [referentialProperty.namespace, ...referentialProperty.namespace.dependencies];
    referentialProperty.mergedProperties.forEach(mergedProperty => {
      failReferencedPropertyDoesNotExist(
        'TargetPropertyPathMustExist',
        namespaces,
        referentialProperty.parentEntity,
        mergedProperty.targetPropertyPath,
        R.head(mergedProperty.mergePropertyPath),
        R.head(mergedProperty.sourceMap.targetPropertyPath),
        failures,
      );
    });
  });

  return failures;
}
