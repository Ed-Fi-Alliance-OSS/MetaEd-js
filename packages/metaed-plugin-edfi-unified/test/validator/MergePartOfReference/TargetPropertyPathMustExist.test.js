// @flow
import {
  newMetaEdEnvironment,
  MetaEdTextBuilder,
  DomainEntityBuilder,
  DomainEntitySubclassBuilder,
  AssociationBuilder,
  AssociationExtensionBuilder,
  AssociationSubclassBuilder,
} from '../../../../../packages/metaed-core/index';
import type { MetaEdEnvironment, ValidationFailure } from '../../../../../packages/metaed-core/index';
import { validate } from '../../../src/validator/MergePartOfReference/TargetPropertyPathMustExist';
import { CommonBuilder } from '../../../../metaed-core/src/builder/CommonBuilder';

describe('when validating domain entity has target property', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName1 = 'DomainEntityName1';
  const domainEntityName2 = 'DomainEntityName2';
  const propertyName = 'Property1';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartDomainEntity(domainEntityName1)
      .withDocumentation('Documentation')
      .withIntegerIdentity(propertyName, 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity(domainEntityName2)
      .withDocumentation('Documentation')
      .withIntegerIdentity(propertyName, 'Documentation')
      .withDomainEntityProperty(domainEntityName1, 'Documentation', false, false)
      .withMergePartOfReference(`${domainEntityName1}.${propertyName}`, `${propertyName}`)
      .withEndDomainEntity()
      .withEndNamespace()
      .sendToListener(new DomainEntityBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build two domain entities', () => {
    expect(metaEd.entity.domainEntity.size).toBe(2);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating domain entity has merge property and entity is wrong', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'Entity1';
  const propertyName = 'Property1';
  const targetEntityName = 'DomainEntity2';
  const targetPropertyName = 'Property2';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartDomainEntity(domainEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity(propertyName, 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity(targetEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity(targetPropertyName, 'Documentation')
      .withDomainEntityProperty(domainEntityName, 'Documentation', false, false)
      .withMergePartOfReference(`${domainEntityName}.${propertyName}`, `EntityNotValid.${targetPropertyName}`)
      .withEndDomainEntity()
      .withEndNamespace()
      .sendToListener(new DomainEntityBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build two domain entities', () => {
    expect(metaEd.entity.domainEntity.size).toBe(2);
  });

  it('should have one validation failure', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('TargetPropertyPathMustExist');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when validating domain entity has merge property and entity is wrong -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when validating domain entity has merge property and entity is wrong -> sourceMap');
  });
});

describe('when validating domain entity has merge property and property is wrong', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'GoodEntity1';
  const propertyName = 'Property1';
  const targetEntityName = 'GoodEntity2';
  const targetPropertyName = 'Property2';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartDomainEntity(domainEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity(propertyName, 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity(targetEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity(targetPropertyName, 'Documentation')
      .withDomainEntityProperty(domainEntityName, 'Documentation', false, false)
      .withMergePartOfReference(`${domainEntityName}.${propertyName}`, `${targetEntityName}.NotThere`)
      .withEndDomainEntity()
      .withEndNamespace()
      .sendToListener(new DomainEntityBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build two domain entities', () => {
    expect(metaEd.entity.domainEntity.size).toBe(2);
  });

  it('should have one validation failure', () => {
    expect(failures).toHaveLength(1);
    expect(failures[0].validatorName).toBe('TargetPropertyPathMustExist');
    expect(failures[0].category).toBe('error');
    expect(failures[0].message).toMatchSnapshot('when validating domain entity has merge property and property is wrong -> message');
    expect(failures[0].sourceMap).toMatchSnapshot('when validating domain entity has merge property and property is wrong -> sourceMap');
  });
});

describe('when validating domain entity has merge property on common type', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'Entity1';
  const secondEntityName = 'SecondEntity2';
  const propertyName = 'Property1';
  const targetEntityName = 'TargetEntity3';
  const targetPropertyName = 'Property2';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartDomainEntity(domainEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity(propertyName, 'Documentation')
      .withEndDomainEntity()

      .withStartCommon(secondEntityName)
      .withDocumentation('Documentation')
      .withDomainEntityIdentity(domainEntityName, 'Documentation')
      .withEndCommon()

      .withStartDomainEntity(targetEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity(targetPropertyName, 'Documentation')
      .withDomainEntityProperty(domainEntityName, 'Documentation', false, false)
      .withCommonProperty(secondEntityName, 'Documentation', false, false)
      .withMergePartOfReference(`${secondEntityName}.${domainEntityName}`, domainEntityName)
      .withEndDomainEntity()
      .withEndNamespace()
      .sendToListener(new DomainEntityBuilder(metaEd, failures))
      .sendToListener(new CommonBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build two domain entities', () => {
    expect(metaEd.entity.domainEntity.size).toBe(2);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating domain entity subclass has merge property', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'EntityWithSub1';
  const propertyName = 'Property1';
  const targetEntityName = 'EntityWithSub2';
  const targetPropertyName = 'Property2';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartDomainEntity(domainEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity('IdentityProperty', 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity(targetEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity('IdentityProperty2', 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntitySubclass('Entity3', targetEntityName)
      .withDocumentation('Documentation')
      .withIntegerProperty(targetPropertyName, 'Documentation', true, false)
      .withDomainEntityProperty(domainEntityName, 'Documentation', false, false)
      .withMergePartOfReference(`${domainEntityName}.${propertyName}`, `${targetPropertyName}`)
      .withEndDomainEntitySubclass()
      .withEndNamespace()
      .sendToListener(new DomainEntityBuilder(metaEd, failures))
      .sendToListener(new DomainEntitySubclassBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build two domain entities', () => {
    expect(metaEd.entity.domainEntity.size).toBe(2);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating association has merge property', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'EntityWithAssoc1';
  const propertyName = 'Property1';
  const targetAssociationName = 'Association';
  const targetPropertyName = 'Property4';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartDomainEntity(domainEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity(propertyName, 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity('Entity2')
      .withDocumentation('Documentation')
      .withIntegerIdentity('Property2', 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity('Entity3')
      .withDocumentation('Documentation')
      .withIntegerIdentity('Property3', 'Documentation')
      .withEndDomainEntity()

      .withStartAssociation(targetAssociationName)
      .withDocumentation('Documentation')
      .withAssociationDomainEntityProperty('Entity2', 'Documentation')
      .withAssociationDomainEntityProperty('Entity3', 'Documentation')
      .withIntegerIdentity(targetPropertyName, 'Documentation')
      .withDomainEntityProperty(domainEntityName, 'Documentation', false, false)
      .withMergePartOfReference(`${domainEntityName}.${propertyName}`, `${targetAssociationName}.${targetPropertyName}`)
      .withEndNamespace()
      .sendToListener(new DomainEntityBuilder(metaEd, failures))
      .sendToListener(new AssociationBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build three domain entities', () => {
    expect(metaEd.entity.domainEntity.size).toBe(3);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating association extension has merge property', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'EntityWithAssocExt1';
  const propertyName = 'Property1';
  const targetAssociationName = 'Association';
  const targetPropertyName = 'Property4';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartDomainEntity(domainEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity(propertyName, 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity('Entity2')
      .withDocumentation('Documentation')
      .withIntegerIdentity('Property2', 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity('Entity3')
      .withDocumentation('Documentation')
      .withIntegerIdentity('Property3', 'Documentation')
      .withEndDomainEntity()

      .withStartAssociation(targetAssociationName)
      .withDocumentation('Documentation')
      .withAssociationDomainEntityProperty('Entity2', 'Documentation')
      .withAssociationDomainEntityProperty('Entity3', 'Documentation')
      .withIntegerIdentity(targetPropertyName, 'Documentation')

      .withStartAssociationExtension(targetAssociationName)
      .withDomainEntityProperty(domainEntityName, 'Documentation', false, false)
      .withMergePartOfReference(`${domainEntityName}.${propertyName}`, `${targetAssociationName}.${targetPropertyName}`)
      .withEndNamespace()
      .sendToListener(new DomainEntityBuilder(metaEd, failures))
      .sendToListener(new AssociationBuilder(metaEd, failures))
      .sendToListener(new AssociationExtensionBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build three domain entities', () => {
    expect(metaEd.entity.domainEntity.size).toBe(3);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating association subclass has merge property', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'EntityWithAssocSub1';
  const propertyName = 'Property1';
  const targetAssociationName = 'Association';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartDomainEntity(domainEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity(propertyName, 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity('Entity2')
      .withDocumentation('Documentation')
      .withIntegerIdentity('Property2', 'Documentation')
      .withEndDomainEntity()

      .withStartDomainEntity('Entity3')
      .withDocumentation('Documentation')
      .withIntegerIdentity('Property3', 'Documentation')
      .withEndDomainEntity()

      .withStartAssociation(targetAssociationName)
      .withDocumentation('Documentation')
      .withAssociationDomainEntityProperty('Entity2', 'Documentation')
      .withAssociationDomainEntityProperty('Entity3', 'Documentation')
      .withIntegerIdentity('Property4', 'Documentation')

      .withStartAssociationSubclass('Entity5', targetAssociationName)
      .withDocumentation('Documentation')
      .withDomainEntityProperty(domainEntityName, 'Documentation', false, false)
      .withMergePartOfReference(`Entity2.${propertyName}`, `${domainEntityName}.${propertyName}`)
      .withEndNamespace()
      .sendToListener(new DomainEntityBuilder(metaEd, failures))
      .sendToListener(new AssociationBuilder(metaEd, failures))
      .sendToListener(new AssociationSubclassBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build three domain entities', () => {
    expect(metaEd.entity.domainEntity.size).toBe(3);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});

describe('when validating abstract entity has merge property', () => {
  const metaEd: MetaEdEnvironment = newMetaEdEnvironment();
  const domainEntityName = 'AbstractEntity1';
  const propertyName = 'Property1';
  const targetEntityName = 'AbstractEntity2';
  const targetPropertyName = 'Property2';
  let failures: Array<ValidationFailure>;

  beforeAll(() => {
    MetaEdTextBuilder.build()
      .withBeginNamespace('edfi')
      .withStartDomainEntity(domainEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity(propertyName, 'Documentation')
      .withEndDomainEntity()

      .withStartAbstractEntity(targetEntityName)
      .withDocumentation('Documentation')
      .withIntegerIdentity(targetPropertyName, 'Documentation')
      .withDomainEntityProperty(domainEntityName, 'Documentation', false, false)
      .withMergePartOfReference(`${domainEntityName}.${propertyName}`, `${targetPropertyName}`)
      .withEndAbstractEntity()
      .withEndNamespace()
      .sendToListener(new DomainEntityBuilder(metaEd, failures));

    failures = validate(metaEd);
  });

  it('should build two domain entities', () => {
    expect(metaEd.entity.domainEntity.size).toBe(2);
  });

  it('should have no validation failures', () => {
    expect(failures).toHaveLength(0);
  });
});
