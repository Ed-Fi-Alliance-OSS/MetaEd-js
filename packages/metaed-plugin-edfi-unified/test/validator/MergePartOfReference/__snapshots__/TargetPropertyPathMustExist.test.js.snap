// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when validating domain entity has merge property across namespaces and entity is wrong should have one validation failure 1`] = `"Merge statement EntityNotValid.Property2 must be a valid path. No Property Property2 was found"`;

exports[`when validating domain entity has merge property across namespaces and entity is wrong should have one validation failure 2`] = `
Object {
  "column": 33,
  "line": 22,
  "tokenText": "EntityNotValid",
}
`;

exports[`when validating domain entity has merge property and entity is wrong should have one validation failure: when validating domain entity has merge property and entity is wrong -> message 1`] = `"Merge statement EntityNotValid.Property2 must be a valid path. No Property Property2 was found"`;

exports[`when validating domain entity has merge property and entity is wrong should have one validation failure: when validating domain entity has merge property and entity is wrong -> sourceMap 1`] = `
Object {
  "column": 33,
  "line": 20,
  "tokenText": "EntityNotValid",
}
`;

exports[`when validating domain entity has merge property and property is wrong should have one validation failure: when validating domain entity has merge property and property is wrong -> message 1`] = `"Merge statement GoodEntity2.NotThere must be a valid path. No Property NotThere was found"`;

exports[`when validating domain entity has merge property and property is wrong should have one validation failure: when validating domain entity has merge property and property is wrong -> sourceMap 1`] = `
Object {
  "column": 37,
  "line": 20,
  "tokenText": "GoodEntity2",
}
`;
